recovery.go

```diff
@@ -7,15 +7,129 @@ import (
 	"os"
 	"runtime"
 	"runtime/debug"
+	"text/template"
 )
 
+const (
+	panicText = "PANIC: %s\n%s"
+	panicHTML = `<html>
+<head><title>PANIC: {{.RecoveredElement}}</title></head>
+<style type="text/css">
+html, body {
+	font-family: Helvetica, Arial, Sans;
+	color: #333333;
+	background-color: #ffffff;
+	margin: 0px;
+}
+h1 {
+	color: #ffffff;
+	background-color: #f14c4c;
+	padding: 20px;
+	border-bottom: 1px solid #2b3848;
+}
+.block {
+	margin: 2em;
+}
+.panic-interface {
+}
+
+.panic-stack-raw pre {
+	padding: 1em;
+	background: #f6f8fa;
+	border: dashed 1px;
+}
+.panic-interface-title {
+	font-weight: bold;
+}
+</style>
+<body>
+<h1>Negroni - PANIC</h1>
+
+<div class="panic-interface block">
+	<h3>{{.RequestDescription}}</h3>
+	<span class="panic-interface-title">Runtime error:</span> <span class="panic-interface-element">{{.RecoveredElement}}</span>
+</div>
+
+{{ if .Stack }}
+<div class="panic-stack-raw block">
+	<h3>Runtime Stack</h3>
+	<pre>{{.StackAsString}}</pre>
+</div>
+{{ end }}
+
+</body>
+</html>`
+	nilRequestMessage = "Request is nil"
+)
+
+var panicHTMLTemplate = template.Must(template.New("PanicPage").Parse(panicHTML))
+
+// PanicInformations contains all
+// elements for printing stack informations.
+type PanicInformations struct {
+	RecoveredElement interface{}
+	Stack            []byte
+	Request          *http.Request
+}
+
+// StackAsString returns a printable version of the stack
+func (p *PanicInformations) StackAsString() string {
+	return string(p.Stack)
+}
+
+//RequestDescription returns a printable description of the url
+func (p *PanicInformations) RequestDescription() string {
+
+	if p.Request == nil {
+		return nilRequestMessage
+	}
+
+	var queryOutput string
+	if p.Request.URL.RawQuery != "" {
+		queryOutput = "?" + p.Request.URL.RawQuery
+	}
+	return fmt.Sprintf("%s %s%s", p.Request.Method, p.Request.URL.Path, queryOutput)
+}
+
+// PanicFormatter is an interface on object can implement
+// to be able to output the stack trace
+type PanicFormatter interface {
+	// ServeStackTrace output the stack for the given answer/response.
+	// If the stack should not be printed, stack will equals "".
+	FormatPanicError(rw http.ResponseWriter, r *http.Request, infos *PanicInformations)
+}
+
+// TextPanicFormatter output the stack
+// as simple text on os.Stdout. If no `Content-Type` is set,
+// it will output the data as `text/plain; charset=utf-8`.
+// Otherwise, the origin `Content-Type` is kept.
+type TextPanicFormatter struct{}
+
+func (t *TextPanicFormatter) FormatPanicError(rw http.ResponseWriter, r *http.Request, infos *PanicInformations) {
+	if rw.Header().Get("Content-Type") == "" {
```

Comment:
Why do we check this here, but not in the `HTMLPanicFormatter`?

Reply:
This was a big question for me. I performed this check inside the `TextPanicFormatter`for backwards compatibility. Now, I was wondering: if you explicetly set up the `HTMLPanicFormatter`, I guessed it was for kind of debugging purposes and I assumed you would like to have it for every potential crash.

But it makes also senses to have the same behaviour as the text version. 

I would propose to mirror the  behaviour of the `TextPanicFormatter` for now and think about a more generic formatter that would handle the output regarding the current `Content-Type` later.

Reply:
:+1: I see, yeah I think it's reasonable to mirror for now.